From 02705f97ee083d72ee13485c485b4f1e12e579af Mon Sep 17 00:00:00 2001
From: Dave Young <dyoung@redhat.com>
Date: Thu, 22 May 2014 02:43:10 -0400
Subject: [firmware] efi: Fix off-by-one bug in EFI Boot Services reservation

Message-id: <1400726607-13717-19-git-send-email-dyoung@redhat.com>
Patchwork-id: 80119
O-Subject: [RHEL7.1 PATCH 18/35] x86/efi: Fix off-by-one bug in EFI Boot Services reservation
Bugzilla: 1080109
RH-Acked-by: Chao WANG <chaowang@redhat.com>
RH-Acked-by: Vivek Goyal <vgoyal@redhat.com>

Resolves: bz1080109

Brew build:
https://brewweb.devel.redhat.com/taskinfo?taskID=7480539

Test: Tested kexec/kdump on HP workstation and qemu.

Upstream commit:
commit a7f84f03f660d93574ac88835d056c0d6468aebe
Author: Dave Young <dyoung@redhat.com>
Date:   Fri Dec 20 18:02:15 2013 +0800

    x86/efi: Fix off-by-one bug in EFI Boot Services reservation

    Current code check boot service region with kernel text region by:
    start+size >= __pa_symbol(_text)
    The end of the above region should be start + size - 1 instead.

    I see this problem in ovmf + Fedora 19 grub boot:
    text start: 1000000 md start: 800000 md size: 800000

    Signed-off-by: Dave Young <dyoung@redhat.com>
    Acked-by: Borislav Petkov <bp@suse.de>
    Acked-by: Toshi Kani <toshi.kani@hp.com>
    Tested-by: Toshi Kani <toshi.kani@hp.com>
    Cc: stable@vger.kernel.org
    Signed-off-by: Matt Fleming <matt.fleming@intel.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/platform/efi/efi.c b/arch/x86/platform/efi/efi.c
index f8ec4da..15e3b5e 100644
--- a/arch/x86/platform/efi/efi.c
+++ b/arch/x86/platform/efi/efi.c
@@ -438,7 +438,7 @@ void __init efi_reserve_boot_services(void)
    * - Not within any part of the kernel
    * - Not the bios reserved area
   */
-  if ((start+size >= __pa_symbol(_text)
+  if ((start + size > __pa_symbol(_text)
     && start <= __pa_symbol(_end)) ||
    !e820_all_mapped(start, start+size, E820_RAM) ||
    memblock_is_region_reserved(start, size)) {
-- 
1.7.1